lib/semantics/expression.cc

```diff
@@ -1779,20 +1770,10 @@ Expr<SubscriptInteger> ExpressionAnalysisContext::Analyze(TypeCategory category,
           [&](const parser::KindSelector::StarSize &x)
               -> Expr<SubscriptInteger> {
             std::intmax_t size = x.v;
-            if (category == TypeCategory::Complex) {
-              // COMPLEX*16 == COMPLEX(KIND=8)
-              if ((size % 2) == 0 &&
-                  evaluate::IsValidKindOfIntrinsicType(category, size / 2)) {
-                size /= 2;
-              } else {
-                Say("COMPLEX*%jd is not a supported type"_err_en_US, size);
-                size = defaultKind;
-              }
-            } else if (!evaluate::IsValidKindOfIntrinsicType(category, size)) {
-              Say("%s*%jd is not a supported type"_err_en_US,
-                  parser::ToUpperCaseLetters(EnumToString(category)).data(),
-                  size);
+            if (!CheckIntrinsicSize(category, size)) {
               size = defaultKind;
+            } else if (category == TypeCategory::Complex) {
+              size /= 2;
```

Comment:
Maybe add a comment as to where in the standard this adjustment comes from?

Reply:
It doesn't come from the standard.  Types with byte sizes like `COMPLEX*16` are universally implemented extensions that predate the standard's concept of kinds.  While kind codes are processor-dependent and arbitrary, the standard does require that the same kind codes for `REAL` have to work with `COMPLEX` (7.4.3.3).  So `COMPLEX*16` means `COMPLEX(KIND=8)`.